From 82ae149d4d195193558c967cbc618f362b5a6eee Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:35:39 -0400
Subject: [block] ensure that the timer is always added

Message-id: <1402670238-13916-52-git-send-email-snitzer@redhat.com>
Patchwork-id: 83798
O-Subject: [RHEL7.1 PATCH 051/150] block: ensure that the timer is always added
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit c7bca4183f73f9d013ea8ae572528b48b5e1ee42
Author: Jens Axboe <axboe@fb.com>
Date:   Fri May 30 15:41:39 2014 -0600

    block: ensure that the timer is always added

    Commit f793aa537866 relaxed the timer addition a little too much.
    If the timer isn't pending, we always need to add it.

    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-timeout.c b/block/blk-timeout.c
index cca7876..8ba914f 100644
--- a/block/blk-timeout.c
+++ b/block/blk-timeout.c
@@ -212,7 +212,7 @@ void blk_add_timer(struct request *req)
    * modifying the timer because expires for value X
    * will be X + something.
    */
-  if (diff >= HZ / 2)
+  if (!timer_pending(&q->timeout) || (diff >= HZ / 2))
    mod_timer(&q->timeout, expiry);
  }
 
-- 
1.7.1